src/dialer/dial-request.js

```diff
@@ -0,0 +1,77 @@
+'use strict'
+
+const AbortController = require('abort-controller')
+const anySignal = require('any-signal')
+const debug = require('debug')
+const errCode = require('err-code')
+const log = debug('libp2p:dialer:request')
+log.error = debug('libp2p:dialer:request:error')
+const FIFO = require('p-fifo')
+const pAny = require('p-any')
+
+class DialRequest {
+  /**
+   *
+   * @param {object} options
+   * @param {Multiaddr[]} options.addrs
+   * @param {TransportManager} options.transportManager
+   * @param {Dialer} options.dialer
+   */
+  constructor ({
+    addrs,
+    dialAction,
+    dialer
+  }) {
+    this.addrs = addrs
+    this.dialer = dialer
+    this.dialAction = dialAction
+  }
+
+  /**
+   * @async
+   * @param {object} options
+   * @param {AbortSignal} options.signal An AbortController signal
+   * @param {number} options.timeout The max dial time for each request
+   * @returns {Connection}
+   */
+  async run (options) {
+    const tokens = this.dialer.getTokens(this.addrs.length)
+    // If no tokens are available, throw
+    if (tokens.length < 1) {
+      throw errCode(new Error('No dial tokens available'), 'ERR_NO_DIAL_TOKENS')
+    }
+
+    const th = new FIFO()
```

Comment:
As we got rid of the `TokenHolder` class, maybe you can name this tokenHolder now? `th` is really abstract